src/coreclr/src/tools/Common/TypeSystem/Interop/IL/Marshaller.cs

```diff
@@ -1447,31 +1447,44 @@ protected override void TransformManagedToNative(ILCodeStream codeStream)
             if (ShouldBePinned)
             {
                 //
-                // Pin the string and push a pointer to the first character on the stack.
+                // Pin the char& and push a pointer to the first character on the stack.
                 //
-                TypeDesc stringType = Context.GetWellKnownType(WellKnownType.String);
+                TypeDesc charRefType = Context.GetWellKnownType(WellKnownType.Char).MakeByRefType();
 
-                ILLocalVariable vPinnedString = emitter.NewLocal(stringType, true);
-                ILCodeLabel lNullString = emitter.NewCodeLabel();
+                ILLocalVariable vPinnedCharRef = emitter.NewLocal(charRefType, true);
 
-                LoadManagedValue(codeStream);
-                codeStream.EmitStLoc(vPinnedString);
-                codeStream.EmitLdLoc(vPinnedString);
+                ILCodeLabel lNonNullString = emitter.NewCodeLabel();
+                ILCodeLabel lCommonExit = emitter.NewCodeLabel();
 
-                codeStream.Emit(ILOpcode.conv_i);
-                codeStream.Emit(ILOpcode.dup);
+                LoadManagedValue(codeStream);
+                codeStream.Emit(ILOpcode.brtrue, lNonNullString);
 
-                // Marshalling a null string?
-                codeStream.Emit(ILOpcode.brfalse, lNullString);
+                //
+                // Null input case
+                // Don't pin anything - load a zero-value nuint (void*) onto the stack
+                //
+                codeStream.Emit(ILOpcode.ldc_i4_0);
+                codeStream.Emit(ILOpcode.conv_u);
+                codeStream.Emit(ILOpcode.br, lCommonExit);
 
+                //
+                // Non-null input case
+                // Extract the char& from the string, pin it, then convert it to a nuint (void*)
+                //
+                codeStream.EmitLabel(lNonNullString);
+                LoadManagedValue(codeStream);
                 codeStream.Emit(ILOpcode.call, emitter.NewToken(
-                    Context.SystemModule.
-                        GetKnownType("System.Runtime.CompilerServices", "RuntimeHelpers").
-                            GetKnownMethod("get_OffsetToStringData", null)));
-
-                codeStream.Emit(ILOpcode.add);
+                    Context.GetWellKnownType(WellKnownType.String).
+                        GetKnownMethod("GetPinnableReference", null)));
+                codeStream.EmitStLoc(vPinnedCharRef);
+                codeStream.EmitLdLoc(vPinnedCharRef);
+                codeStream.Emit(ILOpcode.conv_u);
```

Comment:
Would there be a downside to moving this conv_u to below lCommonExit?  Then the duplicate one in the other branch could be removed.

Reply:
I guess maybe it's better for the same type to be on the stack prior to the branching maybe?

Reply:
Yeah, the stack should be the same at all possible entry points to the branch. In one case the stack will have an _i4_, and in the other case the stack will have a _char\&_. I was concerned that this would break things.

The codegen pattern here is identical to the codegen pattern Roslyn emits for pattern fixed statements. [See Sharplab.](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoQgZgAIBXAOwGcBDAMxjpLoGE6Abyp1RPeoVJIeKOgAVsTAIJMAJgBEIASQwAKAMoYoigOZ0WASiEixtjtgQxVdXXwAWbKACo6AByN0ALzmFja2osKU4dF0mjq6/lAWANxh4QC+aZlRYmm+xgBubBjcktKEsnF6YB7efkZWgnTZ6UA) Honestly I think some minor optimizations are possible, but I was nervous about straying from what the compiler itself already implements.